src/main/java/com/google/cloud/spanner/pgadapter/wireprotocol/QueryMessage.java

```diff
@@ -42,34 +36,34 @@ public class QueryMessage extends ControlMessage {
   protected static final char IDENTIFIER = 'Q';
   public static final String COPY = "COPY";
 
+  private final ParsedStatement statement;
+  private final CopyStatement copyStatement;
+  private final SimpleQueryStatement simpleQueryStatement;
   private final boolean isCopy;
-  private final IntermediateStatement statement;
 
   public QueryMessage(ConnectionHandler connection) throws Exception {
     super(connection);
-    ParsedStatement parsedStatement = PARSER.parse(Statement.of(this.readAll()));
-    this.isCopy = StatementParser.isCommand(COPY, parsedStatement.getSqlWithoutComments());
+    this.statement = PARSER.parse(Statement.of(this.readAll()));
+    this.isCopy = StatementParser.isCommand(COPY, statement.getSqlWithoutComments());
     if (isCopy) {
-      this.statement =
-          new CopyStatement(connection, connection.getServer().getOptions(), parsedStatement);
-    } else if (!connection.getServer().getOptions().requiresMatcher()) {
-      this.statement =
-          new IntermediateStatement(
-              connection.getServer().getOptions(), parsedStatement, this.connection);
+      this.copyStatement =
+          new CopyStatement(connection, connection.getServer().getOptions(), statement);
+      this.simpleQueryStatement = null;
+      this.connection.addActiveStatement(this.copyStatement);
     } else {
-      this.statement =
-          new MatcherStatement(
```

Comment:
We no longer need a custom class for statements that need regex matching. This is now handled in the `SimpleQueryStatement` class.